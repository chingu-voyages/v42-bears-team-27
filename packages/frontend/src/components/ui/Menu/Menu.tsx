import type { RefAttributes, ForwardRefExoticComponent } from 'react';
import { forwardRef } from 'react';
import type { ThemeUIStyleObject } from 'theme-ui';
import * as DropdownMenuPrimitive from '@radix-ui/react-dropdown-menu';
import { MdCheck, MdCircle } from 'react-icons/md';

export interface MenuProps extends DropdownMenuPrimitive.DropdownMenuProps {
  /**
   * Icon to display for button
   */
  icon: React.ReactNode;
  sx?: ThemeUIStyleObject;
  ariaLabel?: string;
}

export const Menu: React.FC<MenuProps> = ({
  children,
  icon,
  sx,
  ariaLabel,
  ...rest
}) => (
  <DropdownMenuPrimitive.Root {...rest}>
    <DropdownMenuPrimitive.Trigger
      aria-label={ariaLabel}
      sx={{ p: 0, bg: 'transparent', border: 'none', ...sx }}
    >
      <div
        sx={{
          variant: 'text.label',
          width: 'min-content',
          bg: 'transparent',
          border: 'none',
          cursor: 'pointer',
          ...sx,
        }}
      >
        {icon}
      </div>
    </DropdownMenuPrimitive.Trigger>
    <DropdownMenuPrimitive.Portal>{children}</DropdownMenuPrimitive.Portal>
  </DropdownMenuPrimitive.Root>
);

if (process.env.NODE_ENV !== 'production') {
  Menu.displayName = 'Menu';
}

export interface MenuContentProps
  extends DropdownMenuPrimitive.DropdownMenuContentProps {
  sx?: ThemeUIStyleObject;
}

export const MenuContent: ForwardRefExoticComponent<
MenuContentProps & RefAttributes<HTMLDivElement>
> = forwardRef(({ children, sx, ...rest }, ref) => (
  <DropdownMenuPrimitive.Content
    ref={ref}
    sx={{
      variant: 'text.label',
      zIndex: 1001,
      minWidth: 220,
      p: 2,
      color: 'text',
      bg: 'muted',
      border: '2px solid currentColor',
      borderRadius: 6,
      cursor: 'default',
      '& div[role^="menuitem"]': {
        pl: 4,
        pr: 1,
        color: 'text',
        cursor: 'pointer',
        '&:hover': {
          bg: 'mutedShade',
          color: 'accent',
        },
      },
      ...sx,
    }}
    {...rest}
  >
    {children}
    <DropdownMenuPrimitive.Arrow />
  </DropdownMenuPrimitive.Content>
));

if (process.env.NODE_ENV !== 'production') {
  MenuContent.displayName = 'MenuContent';
}

export const MenuLabel = DropdownMenuPrimitive.Label;
export const MenuGroup = DropdownMenuPrimitive.Group;

export interface MenuItemProps
  extends DropdownMenuPrimitive.DropdownMenuCheckboxItemProps {
  sx?: ThemeUIStyleObject;
}

export const MenuItem: ForwardRefExoticComponent<
MenuCheckboxItemProps & RefAttributes<HTMLDivElement>
> = forwardRef(({ children, ...rest }, ref) => (
  <DropdownMenuPrimitive.Item ref={ref} {...rest}>
    {children}
  </DropdownMenuPrimitive.Item>
));

if (process.env.NODE_ENV !== 'production') {
  MenuItem.displayName = 'MenuItem';
}

export interface MenuCheckboxItemProps
  extends DropdownMenuPrimitive.DropdownMenuCheckboxItemProps {
  sx?: ThemeUIStyleObject;
}

export const MenuCheckboxItem: ForwardRefExoticComponent<
MenuCheckboxItemProps & RefAttributes<HTMLDivElement>
> = forwardRef(({ children, sx, ...rest }, ref) => (
  <DropdownMenuPrimitive.CheckboxItem
    ref={ref}
    sx={{ position: 'relative', ...sx }}
    {...rest}
  >
    <DropdownMenuPrimitive.ItemIndicator
      sx={{ position: 'absolute', top: '1.5px', left: 1 }}
    >
      <MdCheck size={15} />
    </DropdownMenuPrimitive.ItemIndicator>
    <span>{children}</span>
  </DropdownMenuPrimitive.CheckboxItem>
));

if (process.env.NODE_ENV !== 'production') {
  MenuCheckboxItem.displayName = 'MenuCheckboxItem';
}

export const MenuRadioGroup = DropdownMenuPrimitive.RadioGroup;

export interface MenuRadioItemProps
  extends DropdownMenuPrimitive.DropdownMenuRadioItemProps {
  sx?: ThemeUIStyleObject;
}

export const MenuRadioItem: ForwardRefExoticComponent<
MenuRadioItemProps & RefAttributes<HTMLDivElement>
> = forwardRef(({ children, sx, ...rest }, ref) => (
  <DropdownMenuPrimitive.RadioItem
    ref={ref}
    sx={{ position: 'relative', ...sx }}
    {...rest}
  >
    <DropdownMenuPrimitive.ItemIndicator
      sx={{ position: 'absolute', top: '-1px', left: '6.5px' }}
    >
      <MdCircle size={8} />
    </DropdownMenuPrimitive.ItemIndicator>
    <span>{children}</span>
  </DropdownMenuPrimitive.RadioItem>
));

if (process.env.NODE_ENV !== 'production') {
  MenuRadioItem.displayName = 'MenuRadioItem';
}

export const MenuSeparator: React.FC = () => (
  <DropdownMenuPrimitive.Separator
    sx={{ width: '100%', height: 1, my: 1, bg: 'gray' }}
  />
);
