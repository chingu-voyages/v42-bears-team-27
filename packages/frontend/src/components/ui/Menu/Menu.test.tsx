import { render } from 'test-utils';
import { MdMenu } from 'react-icons/md';

import { Menu } from './Menu';

describe('<Menu />', () => {
  describe('Snapshots', () => {
    it('should match snapshot', () => {
      const { container } = render(<Menu icon={<MdMenu />} />);
      expect(container).toMatchSnapshot();
    });
  });
});
