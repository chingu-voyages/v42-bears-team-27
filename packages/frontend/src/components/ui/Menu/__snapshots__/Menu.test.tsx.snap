// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu /> Snapshots should match snapshot 1`] = `
.emotion-0 {
  padding: 0;
  background-color: transparent;
  border: none;
}

.emotion-1 {
  font-family: "Nunito",sans-serif;
  font-size: 16px;
  font-weight: 400;
  line-height: 1.5;
  width: -webkit-min-content;
  width: -moz-min-content;
  width: min-content;
  background-color: transparent;
  border: none;
  cursor: pointer;
}

<div>
  <button
    aria-expanded="false"
    aria-haspopup="menu"
    class="emotion-0"
    data-state="closed"
    id="radix-:r0:"
    type="button"
  >
    <div
      class="emotion-1"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
        />
      </svg>
    </div>
  </button>
</div>
`;
